ERCS/erc-7943.md

```diff
@@ -0,0 +1,518 @@
+
+Given the agnostic nature of the standard on the specific base token standard being used, the implementation SHOULD use `tokenId = 0` for [ERC-20](./eip-20.md) based implementations, and `amount = 1` for [ERC-721](./eip-721.md) based implementations on `ForcedTransfer` and `Frozen` events, `ERC7943NotAllowedTransfer` and `ERC7943InsufficientUnfrozenBalance` errors, and `forceTransfer`, `setFrozen`, `getFrozen` and `isTransferAllowed` functions. Integrators MAY decide to not enforce this, however the standard discourages it. This is considered a minor tradeoff for having a unique standard interface for different token standards without overlapping syntaxes.
+
+Implementations of this interface MUST implement the necessary functions of their chosen base standard (e.g., [ERC-20](./eip-20.md), [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md) functionalities) and MUST also restrict access to sensitive functions like `forceTransfer` and `setFrozen` using an appropriate access control mechanism (e.g., `onlyOwner`, Role-Based Access Control). The specific mechanism is NOT mandated by this interface standard.
+
+Integrators MUST ensure their transfer methods (`transfer`, `transferFrom`, `mint`, `burn`, and similar) respect the following validation logic, meaning they MUST revert if any of the following validation logics return `false`:
+- Public transfers (`transfer`, `transferFrom`, `safeTransferFrom`, ...) MUST run `isTransferAllowed` and `isUserAllowed` on `from` and `to` addresses.
+- Minting MUST run `isUserAllowed` on `to` address.
+- Burning SHOULD NOT run `isTransferAllowed` nor `isUserAllowed` on the wallet holding the tokens to burn.
+- Burning CAN run validation to prohibit burning more assets than the unfrozen amount. For example, in public burning functions.
+- At the same time, burning CAN bypass freezing validation, in which case it MUST update the freezing status if required, emitting a `Frozen` event before the `Transfer` one. For example, in permissioned burning functions.
+
+The `ERC7943NotAllowedTransfer` and `ERC7943NotAllowedUser` errors CAN be used as a general revert mechanism whenever internal calls to `isUserAllowed` or `isTransferAllowed` return false. Those MIGHT NOT be used or MIGHT be replaced by more specific errors depending on the custom checks performed inside those calls.
+
+In general, error specificity is prioritized, meaning that broad errors such as `ERC7943NotAllowedTransfer` SHOULD be thrown if more specific ones such as `ERC7943InsufficientUnfrozenBalance` do not apply. The same is true for `ERC7943InsufficientUnfrozenBalance` that SHOULD be triggered when a transfer is attempted from `user` with an `amount` of `tokenId` less or equal than its balance, but greater than its unfrozen balance. But if the amount is greater than the whole balance, unrelated from frozen amount, more specific errors from the base standard SHOULD be used instead, like `ERC20InsufficientBalance`.
```

Comment:
> `ERC20InsufficientBalance`

This isn't defined by ERC-20, so might be a little confusing.

Reply:
:pray: changed for a more general statement